runtime/lua/vim/ui.lua

```diff
@@ -88,11 +88,15 @@ function M.input(opts, on_confirm)
   })
 
   opts = (opts and not vim.tbl_isempty(opts)) and opts or vim.empty_dict()
+  opts = vim.tbl_extend('keep', opts, { cancelreturn = vim.NIL })
+
   local input = vim.fn.input(opts)
-  if #input > 0 then
+  if input == vim.NIL then
+    on_confirm(opts.cancelreturn) -- canceled
+  elseif #input >= 0 then
```

Comment:
This seems always true.

Reply:
Hmm you are right. I think I no longer need the else statement below. A corner case is that if a user specifies `cancelreturn = nil`, it will throw an error. Let me fix this.

Reply:
Oh wait. This will throw an error whenever something that does not have a length and is not `nil` or `vim.NIL` as `cancelreturn`.

Reply:
Fixed in fd54b82 and made test cases more precise.